test/src/test/java/jenkins/model/queue/ItemDeletionTest.java

```diff
@@ -0,0 +1,238 @@
+            super.onLoad(parent, name);
+            buildMixIn.onLoad(parent, name);
+        }
+
+        @Override
+        public TopLevelItemDescriptor getDescriptor() {
+            return (DescriptorImpl) Jenkins.get().getDescriptorOrDie(AsyncJob.class);
+        }
+
+        @Override
+        public Queue.Executable createExecutable() throws IOException {
+            return new AsyncRun(this);
+        }
+
+        @Extension
+        public static class DescriptorImpl extends TopLevelItemDescriptor {
+            @Override
+            public TopLevelItem newInstance(ItemGroup parent, String name) {
+                return new AsyncJob(parent, name);
+            }
+        }
+    }
+
+    public static class AsyncRun extends Run<AsyncJob, AsyncRun> implements LazyBuildMixIn.LazyLoadingRun<AsyncJob, AsyncRun>, Queue.Executable {
+        private transient final LazyBuildMixIn.RunMixIn<AsyncJob, AsyncRun> runMixIn = new LazyBuildMixIn.RunMixIn<AsyncJob, AsyncRun>() {
+            @Override protected AsyncRun asRun() {
+                return AsyncRun.this;
+            }
+        };
+
+        public enum State {
+            NOT_STARTED,
+            BUILDING,
+            COMPLETED,
+        }
+
+        private State state = State.NOT_STARTED;
+        private StreamTaskListener listener;
+        private Thread cancellationTask;
+
+        public AsyncRun(AsyncJob job) throws IOException {
+            super(job);
+        }
+
+        @Override
+        public LazyBuildMixIn.RunMixIn<AsyncJob, AsyncRun> getRunMixIn() {
+            return runMixIn;
+        }
+
+        @Override
+        public boolean isBuilding() {
+            return state != State.COMPLETED;
+        }
+
+        @Override
+        public boolean isInProgress() {
+            return state == State.BUILDING;
+        }
+
+        @Override
+        public void run() throws AsynchronousExecution {
+            state = State.BUILDING;
+            try {
+                Files.createDirectories(this.getLogFile().toPath().getParent());
+                Files.createFile(this.getLogFile().toPath());
+                listener = new StreamTaskListener(this.getLogFile(), StandardCharsets.UTF_8);
+                listener.getLogger().println("Starting " + this);
+                RunListener.fireStarted(this, listener);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            throw new AsynchronousExecution() {
+                @Override
+                public void interrupt(boolean forShutdown) {
+                    if (cancellationTask != null) {
+                        return;
+                    }
+                    listener.getLogger().println("Cancelling " + AsyncRun.this);
+                    cancellationTask = new Thread(() -> {
+                        try {
+                            Thread.sleep(2000);
```

Comment:
A bit awkward. We want this to be long enough that the test fails consistently if the fix is reverted, but with the fix there is no need to wait at all. I didn't see any obvious way to optimize things without more invasive changes to `ItemDeletion`.